core/src/apps/cardano/sign_tx.py

```diff
@@ -153,26 +172,65 @@ def _process_outputs(self) -> None:
         self.change_coins = change_coins
         self.change_derivation_paths = change_derivation_paths
 
-    def _build_input_witnesses(self, tx_aux_hash: bytes) -> List[List[bytes, bytes]]:
-        input_witnesses = []
+    def _build_witness(
+        self, tx_body_hash: bytes, address_path: List[int]
+    ) -> List[bytes]:
+        node = self.keychain.derive(address_path)
+        message = b"\x58\x20" + tx_body_hash
+
+        signature = ed25519.sign_ext(
+            node.private_key(), node.private_key_ext(), message
+        )
+
+        public_key = remove_ed25519_prefix(node.public_key())
+
+        return [public_key, signature]
+
+    def _is_certificate_witness_required(self, certificate_type: int) -> bool:
+        return certificate_type != CardanoCertificateType.STAKE_REGISTRATION
+
+    def _build_witnesses(self, tx_aux_hash: bytes) -> List[List[bytes]]:
+        witnesses = []
         for input in self.inputs:
-            node = self.keychain.derive(input.address_n)
-            message = b"\x58\x20" + tx_aux_hash
+            witness = self._build_witness(tx_aux_hash, input.address_n)
+            witnesses.append(witness)
 
-            signature = ed25519.sign_ext(
-                node.private_key(), node.private_key_ext(), message
-            )
+        for certificate in self.certificates:
+            if not self._is_certificate_witness_required(certificate.type):
+                continue
 
-            public_key = remove_ed25519_prefix(node.public_key())
+            witness = self._build_witness(tx_aux_hash, certificate.path)
+            witnesses.append(witness)
 
-            input_witnesses.append([public_key, signature])
+        # todo: GK - resolve for reward and byron witnesses
 
-        # todo: GK - resolve for certificates and byron witnesses
+        return witnesses
 
-        return input_witnesses
+    def _certificate_type_to_cbor_id(self, certificate_type: int) -> int:
+        if certificate_type == CardanoCertificateType.STAKE_REGISTRATION:
+            return 0
+        elif certificate_type == CardanoCertificateType.STAKE_DEREGISTRATION:
+            return 1
+        elif certificate_type == CardanoCertificateType.STAKE_DELEGATION:
+            return 2
+        else:
+            raise ValueError("Unsupported certificate type '%s'" % certificate_type)
```

Comment:
As I think about it now, these cerficate type ids are given in CDDL spec, so I think we want to use them directly (in CardanoCertificateType) --- then we don't need any conversion function. (Only validation of user input --- we do not allow pool registration etc.)

Reply:
I've changed the values and added validation. I am not 100% if it's the right spot for validation, but it's likely going to change when the file is refactored.